clients/cli/devices.c

```diff
@@ -3203,45 +3219,111 @@ do_device_wifi_connect_network (NmCli *nmc, int argc, char **argv)
+	for (i = 0; i < avail_cons->len; i++) {
+		NMRemoteConnection *avail_con = g_ptr_array_index (avail_cons, i);
+		NMSettingWireless *s_wireless = nm_connection_get_setting_wireless (NM_CONNECTION (avail_con));
+		const GByteArray *bssid_arr = bssid2_arr;
+		const char *id = nm_connection_get_id (NM_CONNECTION (avail_con));
+
+		if (!s_wireless)
+			continue;
+
+		if (con_name) {
+			if (!id || strcmp (id, con_name))
+				continue;
+
+			name_match = TRUE;
+		}
+
+		if (!bssid1_arr) {
+			GBytes *ssid = nm_setting_wireless_get_ssid (s_wireless);
+			gchar *utf8;
 
-	if (con_name || private) {
-		s_con =  (NMSettingConnection *) nm_setting_connection_new ();
-		nm_connection_add_setting (connection, NM_SETTING (s_con));
+			if (!ssid)
+				continue;
+
+			utf8 = nm_utils_ssid_to_utf8 (g_bytes_get_data (ssid, NULL),
+			                              g_bytes_get_size (ssid));
 
-		/* Set user provided connection name */
-		if (con_name)
-			g_object_set (s_con, NM_SETTING_CONNECTION_ID, con_name, NULL);
+			if (!utf8 || strcmp (utf8, param_user))
+				continue;
 
-		/* Connection will only be visible to this user when 'private' is specified */
-		if (private)
-			nm_setting_connection_add_permission (s_con, "user", g_get_user_name (), NULL);
+			g_free (utf8);
+		} else
+			bssid_arr = bssid1_arr;
+
+		if (bssid_arr) {
+			const gchar *con_bssid_str = nm_setting_wireless_get_bssid (s_wireless);
+			GByteArray *con_bssid;
+
+			if (!con_bssid_str)
+				continue;
+
+			con_bssid = nm_utils_hwaddr_atoba (con_bssid_str, ETH_ALEN);
+			if (!con_bssid)
+				continue;
+
+			if (memcmp (bssid_arr->data, con_bssid->data, ETH_ALEN)) {
+				g_byte_array_free (con_bssid, TRUE);
+				continue;
+			}
+
+			g_byte_array_free (con_bssid, TRUE);
+		}
+
+		connection = NM_CONNECTION (avail_con);
+		existing_con = TRUE;
+		break;
+	}
+
+	if (name_match && !existing_con) {
+		g_string_printf (nmc->return_text, _("Error: Connection '%s' exists but properties don't match."), con_name);
+		nmc->return_value = NMC_RESULT_ERROR_NOT_FOUND;
+		goto finish;
 	}
-	if (bssid2_arr || hidden) {
-		s_wifi = (NMSettingWireless *) nm_setting_wireless_new ();
-		nm_connection_add_setting (connection, NM_SETTING (s_wifi));
 
-		/* 'bssid' parameter is used to restrict the connection only to the BSSID */
-		if (bssid2_arr)
-			g_object_set (s_wifi, NM_SETTING_WIRELESS_BSSID, bssid2_arr, NULL);
+	if (!existing_con) {
+		/* If there are some connection data from user, create a connection and
+		 * fill them into proper settings. */
+		if (con_name || private || bssid2_arr || password || hidden)
+			connection = nm_simple_connection_new ();
 
-		/* 'hidden' parameter is used to indicate that SSID is not broadcasted */
-		if (hidden) {
-			GBytes *ssid = g_bytes_new (param_user, strlen (param_user));
+		if (!con_name || private) {
+			s_con =  (NMSettingConnection *) nm_setting_connection_new ();
+			nm_connection_add_setting (connection, NM_SETTING (s_con));
```

Comment:
It seem to me that 'connection' can be uninitialized here.

Reply:
Good catch, the `!con_name` is actually a typo.  Not sure how this happened as this part only has more indent added.